### ISSUE TITLE
Reasoning for property assertion based on class restriction "hasValue"

### ISSUE THREAD

**Issue by constnick** [mdesalvo/RDFSharp#166]
Is it possible to implement this feature?

For example, data:
```
Class: ClassWithValueRestrictions
        SubClassOf: 
            propObj value ItemAny,
            propData value 44
    
Individual: ItemDefinedByClassRestrictions
            Types: 
                ClassWithValueRestrictions
    
Individual: ItemAny
```
Inference:
```
Individual: ItemDefinedByClassRestrictions
   
    Facts:  
	propObj  ItemAny,
	propData  44
```

AFAIK Hermit and Pellet reasoners can do it, Hermit makes only object properties assertions, and Pellet makes both object and data properties.

**Comment by mdesalvo** [comment]
Hi,
can you please state this example in Turtle?

Thanks,
Marco

**Comment by constnick** [comment]
[PropAssertion.zip](https://github.com/mdesalvo/RDFSharp/files/5964981/PropAssertion.zip)

**Comment by mdesalvo** [comment]
Thanks for the Turtle, it has clarified to me how Hermit and Pellet work in this reasoning situation:
_((F1 TYPE C) AND (C SUBCLASSOF R) AND (R TYPE RESTRICTION; ONPROPERTY P; HASVALUE F2)) => (F1 P F2)_

This is candidate for a new reasoning rule which may be integrated in the OWL reasoner.

Regards,
Marco

**Comment by mdesalvo** [comment]
Hi, the feature will be available in the next release (v2.21) as a new OWL-DL reasoning rule.
Here a snippet of its usage:
```
RDFNamespaceRegister.AddNamespace(new RDFNamespace("sf", "http://frede.gat/stuff#"));

RDFGraph graph = RDFGraph.FromFile(RDFModelEnums.RDFFormats.Turtle, Environment.CurrentDirectory + "\\Stuff.PropAssertion.ttl");
RDFOntology ontology = RDFOntology.FromRDFGraph(graph);

RDFOntologyReasoner reasoner = new RDFOntologyReasoner().AddRule(RDFOntologyReasonerRuleset.HasValueEntailment);
RDFOntologyReasonerReport report = reasoner.ApplyToOntology(ref ontology);

//Write down the ontology after reasoning process (include both model and data inferences)
ontology.ToRDFGraph(RDFSemanticsEnums.RDFOntologyInferenceExportBehavior.ModelAndData)
        .ToFile(RDFModelEnums.RDFFormats.Turtle, Environment.CurrentDirectory + "\\Stuff.PropAssertion.OUT.ttl");

//Write down pure inferences
report.ToRDFGraph().ToFile(RDFModelEnums.RDFFormats.Turtle, Environment.CurrentDirectory + "\\Stuff.PropAssertion.INF.ttl");
```

I noticed a little glitch in the @base directive of your input ttl file, which should end with "#" in order to correctly compose Uris:
@base <http://frede.gat/stuff#> .

Regards,
Marco
